datalad/support/annexrepo.py

```diff
@@ -549,44 +552,72 @@ def get_file_key(self, file_):
+                    objects.append(obj)
+                except CommandError:
+                    objects.append('')
+
+        return objects
+
     @normalize_paths
-    def file_has_content(self, files):
+    def file_has_content(self, files, allow_quick=True, batch=True):
         """Check whether files have their content present under annex.
 
         Parameters
         ----------
         files: list of str
             file(s) to check for being actually present.
+        allow_quick: bool, optional
+            allow quick check, based on having a symlink into .git/annex/objects.
+            Works only in non-direct mode (TODO: thin mode)
 
         Returns
         -------
         list of bool
-            Per each input file states either file has content locally
+            For each input file states either file has content locally
         """
         # TODO: Also provide option to look for key instead of path
 
-        try:
-            out, err = self._run_annex_command('find', annex_options=files,
-                                               expect_fail=True)
-        except CommandError as e:
-            if e.code == 1 and "not found" in e.stderr:
-                if len(files) > 1:
-                    lgr.debug("One of the files was not found, so performing "
-                              "'find' operation per each file")
-                    # we need to go file by file since one of them is non
-                    # existent and annex pukes on it
-                    return [self.file_has_content(file_) for file_ in files]
-                return [False]
-            else:
-                raise
-
-        found_files = {f for f in out.splitlines() if f}
-        found_files_new = set(found_files) - set(files)
-        if found_files_new:
-            raise RuntimeError("'annex find' returned entries for files which "
-                               "we did not expect: %s" % (found_files_new,))
-
-        return [file_ in found_files for file_ in files]
+        if self.is_direct_mode() or not allow_quick:  # TODO: thin mode
+            # TODO: Also provide option to look for key instead of path
+            find = self.find(files, normalize_paths=False, batch=batch)
+            return [bool(filename) for filename in find]
+        else:  # ad-hoc check which should be faster than call into annex
+            out = []
+            for f in files:
+                filepath = opj(self.path, f)
+                target_path = ''
+                if islink(filepath):                    # if symlink
+                    target_path = readlink(filepath)    # find link target
+                    # convert to absolute path if not
+                    target_path = target_path \
+                        if isabs(target_path) \
+                        else opj(dirname(filepath), target_path)
+                out.append(exists(target_path) and islink(filepath))
```

Comment:
what if it is a symlink just pointing to another present file which is not even under annex?  I guess you would need to add a check that `realpath(target_path).startswith(opj(self.path, '.git', 'annex', 'objects')))` ?